iina/ViewLayer.swift

```diff
@@ -95,8 +90,8 @@ class ViewLayer: CAOpenGLLayer {
 
   override func draw(inCGLContext ctx: CGLContextObj, pixelFormat pf: CGLPixelFormatObj, forLayerTime t: CFTimeInterval, displayTime ts: UnsafePointer<CVTimeStamp>?) {
     let mpv = videoView.player.mpv!
-    needsMPVRender = false
 
+    CGLLockContext(ctx)
```

Comment:
The calls to `CGLLockContext` and `CGLUnlockContext` **should be** removed from the `draw` method, but not for the reason cited.

OpenGL context must be locked even if the main thread is used exclusively, because the application does not control this. Quartz performs some operations asynchronously that access the context using a separate thread. We noticed this in https://github.com/mpv-player/mpv/issues/11681. See [this comment](https://github.com/mpv-player/mpv/issues/11681#issuecomment-1586568804).

**However** in the case of the `draw` method  the locking should be removed because:

```text
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x00000001fb4334f8 OpenGL`CGLLockContext
    frame #1: 0x000000019de92998 QuartzCore`CAOpenGLLayerDraw(CAOpenGLLayer*, double, CVTimeStamp const*, unsigned int) + 548
    frame #2: 0x000000019de925a4 QuartzCore`-[CAOpenGLLayer _display] + 520
    frame #3: 0x00000001044b7c5c IINA`closure #1 in ViewLayer.display(isUninited=false, self=0x0000600002769400) at ViewLayer.swift:160:13
```

That is `CAOpenGLLayer` locking the context before calling `ViewLayer.draw` as discussed in the documentation for [CAOpenGLLayer.draw](https://developer.apple.com/documentation/quartzcore/caopengllayer/1522316-draw). Wait a minute. The documentation does not say a thing about this. Nothing. Nada. Zip. Zilch. Not one iota. Could it be somebody got nailed by this causing a deadlock and thus previously has had to debug `CAOpenGLLayer` and try and figure out what it was doing…

As `OpenGL` uses a reentrant lock the locking in `draw` works, but adds needless overhead and should be removed.

Reply:
Yes, I agree, the locking in `CAOpenGLLayer` may explain why the `mpv` crash only occurs during shutdown.

The old archived Apple documentation is in may ways significantly better than the current documentation. But as it is way out of date so some details may have changed. And there are definitely problems even in the old documentation. In [this](https://developer.apple.com/library/archive/samplecode/CALayerEssentials/Listings/ExampleCAOpenGLLayer_m.html#//apple_ref/doc/uid/DTS40008029-ExampleCAOpenGLLayer_m-DontLinkElementID_8) `CALayerEssentials` example the comment before the method `drawInCGLContext` says:
> 1. Draw your OpenGL content! (the current context has already been set)

And the method starts with:
> // Set the current context to the one given to us.
> CGLSetCurrentContext(glContext);

This shows one of the reasons why the software industry keeps trying to come up with "easier" ways to support threading. When it comes to threading poor documentation of "black box" implementations can result in severe problems that are hard to reproduce.